src/interpreter/mod.rs

```diff
@@ -734,16 +734,12 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
             ty::TyRef(_, ty::TypeAndMut { ty, .. }) |
             ty::TyRawPtr(ty::TypeAndMut { ty, .. }) |
             ty::TyBox(ty) => {
-                if self.type_is_sized(ty) {
-                    assert_eq!(field_index, 0);
-                    Ok(self.tcx.mk_imm_ptr(self.tcx.types.never))
-                } else {
-                    match (field_index, &ty.sty) {
-                        (1, &ty::TySlice(_)) => Ok(self.tcx.types.usize),
-                        (1, &ty::TyTrait(_)) |
-                        (0, _) => Ok(self.tcx.mk_imm_ptr(self.tcx.types.never)),
-                        _ => bug!("invalid fat pointee type"),
-                    }
+                assert!(!self.type_is_sized(ty));
+                match (field_index, &self.tcx.struct_tail(ty).sty) {
```

Comment:
@eddyb Is this a problem because of the tuple? Not being copyable for sure wouldn't be a problem when matching directly, because the patterns don't try to copy (or move, or bind at all) anything in the `TyS`.